sysl2/sysl/ValueUtil.go

```diff
@@ -0,0 +1,233 @@
+		if typeName == "sequence" {
+			seqMap := MakeValueMap()
+			addItemToValueMap(m.GetMap(), typeName, seqMap)
+
+			typeName, typeDetail = getTypeDetail(t.GetSequence())
+			addItemToValueMap(seqMap.GetMap(), "type", MakeValueString(typeName))
+			addItemToValueMap(seqMap.GetMap(), typeName, MakeValueString(typeDetail))
+		}
+
+		addItemToValueMap(fieldMap.GetMap(), key, m)
+		addItemToValueMap(m.GetMap(), "name", MakeValueString(key))
+		addItemToValueMap(m.GetMap(), "optional", MakeValueBool(t.Opt))
+		addItemToValueMap(m.GetMap(), "docstring", MakeValueString(t.Docstring))
+	}
+	return fieldMap
+}
+
+func typeToValue(t *sysl.Type) *sysl.Value {
+	m := MakeValueMap()
+	addItemToValueMap(m.GetMap(), "docstring", MakeValueString(t.Docstring))
+	addItemToValueMap(m.GetMap(), "attrs", attrsToValueMap(t.Attrs))
+	var typeName string
+	switch x := t.Type.(type) {
+	case *sysl.Type_Tuple_:
+		typeName = "tuple"
+		addItemToValueMap(m.GetMap(), "fields", fieldsToValueMap(x.Tuple.AttrDefs))
+	case *sysl.Type_Relation_:
+		typeName = "relation"
+		addItemToValueMap(m.GetMap(), "fields", fieldsToValueMap(x.Relation.AttrDefs))
+	case *sysl.Type_OneOf_:
+		unionSet := MakeValueSet()
+		typeName = "union"
+		for _, embeddedType := range x.OneOf.Type {
+			appendItemToValueList(unionSet.GetSet(), MakeValueString(embeddedType.GetTypeRef().Ref.Path[0]))
+		}
+		addItemToValueMap(m.GetMap(), "fields", unionSet)
+	default:
+		panic("typeToValue: unsupported type")
+	}
+	addItemToValueMap(m.GetMap(), "type", MakeValueString(typeName))
+	return m
+}
+
+func attrsToValueMap(attrs map[string]*sysl.Attribute) *sysl.Value {
+	m := MakeValueMap()
+	for key, attr := range attrs {
+		addItemToValueMap(m.GetMap(), key, attributeToValue(attr))
+	}
+	return m
+}
+
+func typesToValueMap(types map[string]*sysl.Type) *sysl.Value {
+	m := MakeValueMap()
+	for key, t := range types {
+		switch t.Type.(type) {
+		case *sysl.Type_OneOf_:
+			continue
+		default:
+			addItemToValueMap(m.GetMap(), key, typeToValue(t))
+		}
+	}
+	return m
+}
+
+func unionToValueMap(types map[string]*sysl.Type) *sysl.Value {
+	m := MakeValueMap()
+	for key, t := range types {
+		switch t.Type.(type) {
+		case *sysl.Type_OneOf_:
+			addItemToValueMap(m.GetMap(), key, typeToValue(t))
+		}
+	}
+	return m
+}
+
+// AddInt add int to scope
+func (s *Scope) AddInt(name string, val int64) {
+	(*s)[name] = MakeValueI64(val)
+}
+
+// AddString add string to scope
+func (s *Scope) AddString(name string, val string) {
+	(*s)[name] = MakeValueString(val)
+}
+
+// AddApp add sysl.App to scope
+func (s *Scope) AddApp(name string, app *sysl.Application) {
+	m := MakeValueMap()
+	(*s)[name] = m
+	addItemToValueMap(m.GetMap(), "name", MakeValueString(getAppName(app.Name)))
+	addItemToValueMap(m.GetMap(), "attrs", attrsToValueMap(app.Attrs))
+	addItemToValueMap(m.GetMap(), "types", typesToValueMap(app.Types))
+	addItemToValueMap(m.GetMap(), "union", unionToValueMap(app.Types))
```

Comment:
This is a good example of where a parameter to `MakeValueMap()` would really clean things up:
```go
	(*s)[name] = MakeValueMap({
		"name": MakeValueString(getAppName(app.Name)),
		"attrs": attrsToValueMap(app.Attrs),
		"types": typesToValueMap(app.Types),
		"union": unionToValueMap(app.Types),
	})
```